### PR TITLE
Make Ozone as a under file system of Alluxio

### LINKED ISSUE #11390: Ozone as a under file system of Alluxio
**Is your feature request related to a problem? Please describe.**

Ozone is a next generation distributed storage for big data ecosystem. Ozone is a scalable, redundant, and distributed object store for Hadoop. Apart from scaling to billions of objects of varying sizes, Ozone can function effectively in containerized environments such as Kubernetes and YARN.

If Alluxio support ozone as a UFS, Alluxio can be a cache and bring ozone into global namesystem.

**Describe the solution you'd like**

Support ozone as UNDERFS through develop a module of Alluxio

**Describe alternatives you've considered**
develop a module under underfs module, named alluxio-underfs-ozone

**Urgency**
Normal

**Additional context**
No

### PR DESCRIPTION
**Pull request by maobaolong**
This PR introduces a new top-level UFS module for compiling an Ozone-compatible UFS. The Ozone client is able to use the Hadoop FileSystem interface, so we don't need to write any new logic to support Ozone.

The reason we include this as a top-level UFS as opposed to including the ozone libraries directly in the HDFS UFS is to gain more control over the supported ozone version as well as reduce the size of the release tarball. Otherwise, we would maintain many copies of the same Ozone jars.

Resolves #11390

### CODE REVIEW

**Comment by maobaolong** [comment]
image

I think the failed test isn't caused by this PR.

File: pom.xml [hunk 1]
```diff
@@ -151,6 +151,7 @@
     <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
     <slf4j.version>1.7.2</slf4j.version>
     <jackson.version>2.10.1</jackson.version>
+    <ozone.version>0.5.0-beta</ozone.version>
```
Comment (ZacBlanco) [inline pom.xml:154]: Any reason specifically to use the beta? I would prefer to have official releases.

File: underfs/ozone/src/main/java/alluxio/underfs/ozone/OzoneUnderFileSystemFactory.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+    if (checkOzoneCredentials(conf)) {
+      try {
+        return OzoneUnderFileSystem.createInstance(new AlluxioURI(path), conf);
+      } catch (Exception e) {
+        throw Throwables.propagate(e);
+      }
+    }
+
+    String err = "COS Credentials not available, cannot create COS Under File System.";
```
Comment (ZacBlanco) [inline underfs/ozone/src/main/java/alluxio/underfs/ozone/OzoneUnderFileSystemFactory.java:46]: This should be `Ozone`?

File: underfs/ozone/src/main/java/alluxio/underfs/ozone/OzoneUnderFileSystem.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+   * @return a new HDFS {@link UnderFileSystem} instance
+   */
+  public static OzoneUnderFileSystem createInstance(AlluxioURI ufsUri,
+      UnderFileSystemConfiguration conf) {
+    Configuration hdfsConf = createConfiguration(conf);
+    return new OzoneUnderFileSystem(ufsUri, conf, hdfsConf);
+  }
+
+  /**
+   * Constructs a new HDFS {@link UnderFileSystem}.
```
Comment (ZacBlanco) [inline underfs/ozone/src/main/java/alluxio/underfs/ozone/OzoneUnderFileSystem.java:43]: Should this say `Ozone` or `O3FS`?

File: underfs/ozone/src/main/java/alluxio/underfs/ozone/OzoneUnderFileSystem.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+   *
+   * @param ufsUri the {@link AlluxioURI} for this UFS
+   * @param conf the configuration for this UFS
+   * @param hdfsConf the configuration for HDFS
```
Comment (ZacBlanco) [inline underfs/ozone/src/main/java/alluxio/underfs/ozone/OzoneUnderFileSystem.java:47]: should this say `Ozone` or `O3FS`?

File: underfs/ozone/src/main/java/alluxio/underfs/ozone/OzoneUnderFileSystemFactory.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+  /**
+   * @param conf optional configuration object for the UFS
+   *
+   * @return true if both access, secret and endpoint keys are present, false otherwise
+   */
+  private boolean checkOzoneCredentials(UnderFileSystemConfiguration conf) {
+    // TODO(maobaolong): impl check logic while ozone supported
+    return true;
+  }
```
Comment (ZacBlanco) [inline underfs/ozone/src/main/java/alluxio/underfs/ozone/OzoneUnderFileSystemFactory.java:63]: is there a reason this isn't implemented yet?

File: webui/npm-shrinkwrap.json [hunk 6]
```diff
@@ -2055,36 +2055,11 @@
         "lodash": "^4.17.4"
       }
     },
-    "@nodelib/fs.scandir": {
```
Comment (ZacBlanco) [inline webui/npm-shrinkwrap.json:2058]: Can you revert the changes to this file?

Comment (maobaolong) [reply webui/npm-shrinkwrap.json]: Sorry, my fault

**Comment by maobaolong** [comment]
@ZacBlanco  Thanks for your tips, you make my PR more clear with the following config.

```properties
alluxio.underfs.hdfs.prefixes=hdfs://,glusterfs:///,o3fs://
```

It can work well using the following cli command of alluxio to mount o3fs

```bash
 bin/alluxio fs mount  --option alluxio.underfs.hdfs.configuration=/Users/mbl/projects/hadoop-ozone/hadoop-ozone/dist/target/ozone-0.6.0-SNAPSHOT/etc/hadoop/ozone-site.xml:/Users/mbl/projects/hadoop-ozone/hadoop-ozone/dist/target/ozone-0.6.0-SNAPSHOT/etc/hadoop/core-site.xml  /ozone o3fs://mybucket.myvol/

```

File: underfs/ozone/src/main/java/alluxio/underfs/ozone/OzoneUnderFileSystem.java (hunk 4 above)
Comment (maobaolong) [reply underfs/ozone/src/main/java/alluxio/underfs/ozone/OzoneUnderFileSystem.java]: I remove these class.

File: underfs/ozone/src/main/java/alluxio/underfs/ozone/OzoneUnderFileSystem.java (hunk 3 above)
Comment (maobaolong) [reply underfs/ozone/src/main/java/alluxio/underfs/ozone/OzoneUnderFileSystem.java]: I remove these class.

File: underfs/ozone/src/main/java/alluxio/underfs/ozone/OzoneUnderFileSystemFactory.java (hunk 5 above)
Comment (maobaolong) [reply underfs/ozone/src/main/java/alluxio/underfs/ozone/OzoneUnderFileSystemFactory.java:37]: I let this class extents HdfsUnderFileSystemFactory, so the only thing i should override is method 'supportsPath'

**Comment by maobaolong** [comment]
@ZacBlanco After discuss with you, i think make ozone as top-level ufs module is better, the newest PR can success work with ozone now

File: underfs/ozone/pom.xml [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+  <modelVersion>4.0.0</modelVersion>
+  <parent>
+    <artifactId>alluxio-underfs</artifactId>
+    <groupId>org.alluxio</groupId>
+    <version>2.3.0-SNAPSHOT</version>
+  </parent>
+
+  <artifactId>alluxio-underfs-ozone</artifactId>
+  <name>Alluxio Under File System - Apache ozone</name>
+  <description>Apache ozone Under File System implementation</description>
```
Comment (ZacBlanco) [inline underfs/ozone/pom.xml:23]:
```suggestion
  <description>Apache Ozone Under File System implementation</description>
```

File: underfs/ozone/pom.xml [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+  <modelVersion>4.0.0</modelVersion>
+  <parent>
+    <artifactId>alluxio-underfs</artifactId>
+    <groupId>org.alluxio</groupId>
+    <version>2.3.0-SNAPSHOT</version>
+  </parent>
+
+  <artifactId>alluxio-underfs-ozone</artifactId>
+  <name>Alluxio Under File System - Apache ozone</name>
```
Comment (ZacBlanco) [inline underfs/ozone/pom.xml:22]:
```suggestion
  <name>Alluxio Under File System - Apache Ozone</name>
```

File: underfs/ozone/src/main/java/alluxio/underfs/ozone/OzoneUnderFileSystemFactory.java [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+import alluxio.underfs.UnderFileSystemConfiguration;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * Factory for creating {@link HdfsUnderFileSystem}.
+ *
+ * It caches created {@link HdfsUnderFileSystem}s, using the scheme and authority pair as the key.
+ */
+@ThreadSafe
```
Comment (ZacBlanco) [inline underfs/ozone/src/main/java/alluxio/underfs/ozone/OzoneUnderFileSystemFactory.java:26]: Can you update the javadoc to reference `OzoneUnderFileSystem` instead of `HdfsUnderFileSystem`?

File: pom.xml (hunk 1 above)
Comment (maobaolong) [reply pom.xml:154]: Because for ozone, ozone will not have release version recently, and the beta version is the most stable version now, reference **https://hadoop.apache.org/ozone/downloads/**

File: underfs/ozone/src/main/java/alluxio/underfs/ozone/OzoneUnderFileSystemFactory.java (hunk 9 above)
Comment (maobaolong) [reply underfs/ozone/src/main/java/alluxio/underfs/ozone/OzoneUnderFileSystemFactory.java:26]: @ZacBlanco I have no OzoneUnderFileSystem now, it is removed, i use HdfsUnderFileSystem for access ozone

**Comment by maobaolong** [comment]
@ZacBlanco Thank you for your suggestion, i accept all your commit, PTAL.

**Comment by ZacBlanco** [comment]
Adding a note that @maobaolong tested adding the Ozone jars directly to the Alluxio classpath with the `ALLUXIO_CLASSPATH` environment variable and setting `alluxio.underfs.hdfs.prefixes` to include `o3fs://`.

Unfortunately, this didn't work so it required the code changes in this PR.

**Comment by ZacBlanco** [comment]
@apc999 would you like to review this?

**Comment by apc999** [comment]
alluxio-bot, merge this please

**Comment by maobaolong** [comment]
@ZacBlanco @apc999  Thanks for your kindly review, i view write a docs later. The following track this task. 

https://github.com/Alluxio/alluxio/issues/11431
